upstairs/src/lib.rs

```diff
@@ -5056,10 +5056,43 @@ impl Upstairs {
          * Construct a list of dependencies for this write based on the
          * following rules:
          *
-         * - ignore everything that happened before the last flush
          * - writes have to depend on the last flush completing
          * - any overlap of impacted blocks requires a dependency
          *
+         * It's important to remember that jobs may arrive at different
+         * Downstairs in different orders (they should still complete in job
+         * dependency order!). For example, say that searching for the
+         * dependency of a write stopped at the last flush. Then say that the
+         * following set of jobs were submitted:
+         *
+         *       block
+         * op# | 0 1 2 | deps
+         * ----|-------------
+         *   0 | R R   |
+         *   1 | F F F |
+         *   2 |   W W | 1
+         *
+         * Without any dependencies, a downstairs could choose to perform op 0
+         * at any time, including after the write! This would result in an
+         * incorrect read. It's important to search for write dependencies in
+         * the list of all active jobs.
+         *
+         * Take another example. Again say that a flush was a barrier to
+         * searching for write dependencies:
+         *
+         *       block
+         * op# | 0 1 2 | deps
+         * ----|-------------
+         *   0 | W     |
+         *   1 |   W   |
+         *   2 | FFFFF | 0,1
+         *   3 | W     | 2
+         *   4 |   W   | 2
+         *
+         * Without dependencies between ops 0 and 3, and 1 and 4, there's
+         * nothing that prevents an aggressively parallel downstairs from
+         * executing these writes in the wrong order.
```

Comment:
I think this case is actually okay. 
Job 3 depends on 2,
Job 2 depends on 0,1

So, job 3 can't go untill 0 and 1 (and then 2) are all done.

Reply:
Yeah, this part of the comment is wrong. Removed in a0b0b2d